src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Utf16StringMarshaller.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Marshaller for UTF-16 strings
+    /// </summary>
+    [CLSCompliant(false)]
+    [CustomTypeMarshaller(typeof(string), BufferSize = 0x100,
+        Features = CustomTypeMarshallerFeatures.UnmanagedResources | CustomTypeMarshallerFeatures.TwoStageMarshalling | CustomTypeMarshallerFeatures.CallerAllocatedBuffer)]
+    public unsafe ref struct Utf16StringMarshaller
+    {
+        private ushort* allocated;
+        private Span<ushort> span;
+        private bool isNullString;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Utf16StringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        public Utf16StringMarshaller(string str)
+            : this(str, default(Span<ushort>))
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Utf16StringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        /// <param name="buffer">Buffer that may be used for marshalling.</param>
+        /// <remarks>
+        /// <seealso cref="CustomTypeMarshallerFeatures.CallerAllocatedBuffer"/>
+        /// </remarks>
+        public Utf16StringMarshaller(string str, Span<ushort> buffer)
+        {
+            isNullString = false;
+            span = default;
+            if (str is null)
+            {
+                allocated = null;
+                isNullString = true;
+            }
+            else if ((str.Length + 1) < buffer.Length)
+            {
+                span = buffer;
+                str.AsSpan().CopyTo(MemoryMarshal.Cast<ushort, char>(buffer));
+                // Supplied memory is in an undefined state so ensure
+                // there is a trailing null in the buffer.
+                span[str.Length] = '\0';
```

Comment:
For in-only string arguments, do we not support just pinning the string itself and passing the pinned pointer to native, rather than copying it to the stack or newly-allocated memory?

Reply:
The generated code does just pin for by-value (so it doesn't actually create/use this marshaller). For the implementation of the marshaller itself, we made it copy so that it has the logic of in only means in only, rather than implicitly in/out.